### PR TITLE
Allow ethereum permissions on non secure origins

### PR DESCRIPTION
**Pull request by bbondy**
Crypto Wallets and other sites allow HTTP based Ethereum permission prompts. We should have this for parity. It is common for development for example to use localhost over HTTP especially.

Resolves https://github.com/brave/brave-browser/issues/18143

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/permissions/contexts/brave_ethereum_permission_context.cc [hunk 1]
```diff
@@ -47,7 +47,7 @@ BraveEthereumPermissionContext::BraveEthereumPermissionContext(
 BraveEthereumPermissionContext::~BraveEthereumPermissionContext() = default;
 
 bool BraveEthereumPermissionContext::IsRestrictedToSecureOrigins() const {
-  return true;
+  return false;
```
Comment (yrliou) [inline components/permissions/contexts/brave_ethereum_permission_context.cc:50]: Can you add the descriptions you have in the PR of why we need this as a comment here please? Thanks.

**Review by yrliou — APPROVED** [review]

File: components/permissions/contexts/brave_ethereum_permission_context.cc (hunk 1 above)
Comment (bbondy) [reply components/permissions/contexts/brave_ethereum_permission_context.cc:53]: fixed

File: components/permissions/contexts/brave_ethereum_permission_context.cc [hunk 2]
```diff
@@ -47,7 +47,10 @@ BraveEthereumPermissionContext::BraveEthereumPermissionContext(
 BraveEthereumPermissionContext::~BraveEthereumPermissionContext() = default;
 
 bool BraveEthereumPermissionContext::IsRestrictedToSecureOrigins() const {
-  return true;
+  // For parity with Crypto Wallets and MM we should allow a permission prompt
+  // to be shown for HTTP sites. Developers often use localhost for development
+  // for example.
```
Comment (diracdeltas) [inline components/permissions/contexts/brave_ethereum_permission_context.cc:52]: FYI i think you can allow localhost without allowing all HTTP. according to the secure contexts spec, localhost should be considered secure. https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts

are there other HTTP sites that require this functionality?

Comment (bbondy) [reply components/permissions/contexts/brave_ethereum_permission_context.cc:52]: Our current wallet Crypto Wallets and MM both expose window.ethereum to HTTP sites, not only localhost. 
So this is for webcompat / parity with that.
I don't know of which dapps are popular or not, but it's very feasible that users use ganache or a node on their lan, or use a dapp that is only on HTTP.

Comment (bbondy) [reply components/permissions/contexts/brave_ethereum_permission_context.cc:52]: Just to add a bit more info...
`window.ethereum` will be exposed wether or not a site is HTTP before and after this PR both. Also a site can use `window.ethereum` (including making requests to ethereum over HTTPS even if the site is HTTP) without permissions.

This PR allows an HTTP site to request permissions though.
The permission gives the site access to:
- The selected user allowed address
- Adding and signing transactions for that address (requires user approval still)
- Signing messages for that address (requires user approval still)

**Comment by bbondy** [comment]
Talked to Yan in Slack and she is ok with me landing this. So merging.

**Comment by bbondy** [comment]
> Talked to Yan in Slack and she is ok with me landing this. So merging.

Just saw this now but I answered this recently here:
https://github.com/brave/brave-core/pull/10105/files#r711043885

**Pull request merged**